Perf events metrics

And42 opened the pull request:
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

Changelog category (leave one):
- New Feature


Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
- Performance metrics via perf_events (Linux only)


Detailed description / Documentation draft:
Using perf_events on Linux for performance metrics. All the metrics are query based. So, you can find them all within `system.query_log` table. Each metric opens a separate file descriptor.

Tasks:
- [x] ~~Come up with additional _meaningful_ metrics based on the raw ones~~ *(so many raw ones, additional can be added in the future if needed)*
- [x] Remove metrics that already exist in other sources
- [x] _Probably_ lower number of file descriptors per query (~event groups could help?~ - *could not*). Overall, large number of [file descriptors](https://github.com/ClickHouse/ClickHouse/pull/9545#issuecomment-620468513)
- [x] Add `CAP_SYS_ADMIN` check
- [x] Rename metrics to some the user is familiar with. (~~e.g. `perf stat` names as advised [here](https://github.com/ClickHouse/ClickHouse/pull/9545#issuecomment-620451611)~~; *camel case shortened now*)
- [x] [~~Fix multiple `PerfEventsCounters` creation~~](https://github.com/ClickHouse/ClickHouse/pull/9545#issuecomment-618593158) *(overcomplicates the logic and not needed at the moment)*
- [x] [Fix invalid counter measurements](https://github.com/ClickHouse/ClickHouse/pull/9545#issuecomment-620545743) *(removed counters)*

Requirements
---
- `perf_events`
- `CAP_SYS_ADMIN` capability for an executable or `perf_event_paranoid` value lower than 3:
  - `perf_event_paranoid` >=1: CPU events info unavailable
  - `perf_event_paranoid` >=2: kernel profiling unavailable

Lifecycle
---
Query thread started > `initializeProfileEvents` called > Start measuring > Query thread stops > `finalizeProfileEvents` called > Increment counters + hold descriptors for the future events on the same thread

There are no intermediate counters' updates between _Query thread started_ and _Query thread stops_ as all the counters have an initial value of 0 unlike `RUsageCounters` and `TasksStatsCounters` and if an overflow happens nothing (as of now) can be done to handle it somehow properly. But an overflow is unlikely to happen as all events are stored as 64-bit unsigned integers and it will take `2^64 / 5000000000 / 60 / 60 / 24 / 365 ~= 117` years of a non-stop query on the 5 GHz CPU to overflow the fastest growing `instructions` counter.

Messages Handling
---
All the warnings (e.g. when there are issues while working with file descriptors) are written in `LOG_WARNING`.

Hardware Events
-----

*NB:* each hardware event has two additional profile events associated: "<event>Running" and "<event>Enabled" that contain times when measurement was running on the PMU and enabled. This can be useful to get an idea of a measurement's precision as a CPU has a limited number of PMU's and it technically can't measure more than "number of PMU's" events at the same time. 

| Name | Settings Alias | Description |
| --- | --- | --- |
| PerfCpuCycles | cpu-cycles | Total cycles. Be wary of what happens during CPU frequency scaling. |
| PerfInstructions | instructions | Retired instructions. Be careful, these can be affected by various issues, most notably hardware interrupt counts. |
| PerfCacheReferences | cache-references | Cache accesses. Usually this indicates Last Level Cache accesses but this may vary depending on your CPU. This may include prefetches and coherency messages; again this depends on the design of your CPU. |
| PerfCacheMisses | cache-misses | Cache misses. Usually this indicates Last Level Cache misses; this is intended to be used in con‐junction with the PERF_COUNT_HW_CACHE_REFERENCES event to calculate cache miss rates. |
| PerfBranchInstructions | branch-instructions | Retired branch instructions. Prior to Linux 2.6.35, this used the wrong event on AMD processors. |
| PerfBranchMisses | branch-misses | Mispredicted branch instructions. |
| PerfBusCycles | bus-cycles | Bus cycles, which can be different from total cycles. |
| PerfStalledCyclesFrontend | stalled-cycles-frontend | Stalled cycles during issue. |
| PerfStalledCyclesBackend | stalled-cycles-backend | Stalled cycles during retirement. |
| PerfRefCpuCycles | ref-cpu-cycles | Total cycles; not affected by CPU frequency scaling. |

Software Events
---
| Name | Settings Alias | Description |
| --- | --- | --- |
| PerfCpuClock | cpu-clock | The CPU clock, a high-resolution per-CPU timer |
| PerfTaskClock | task-clock | A clock count specific to the task that is running |
| ~~PerfPageFaults~~ | page-faults | Number of page faults |
| PerfContextSwitches | context-switches | Number of context switches |
| PerfCpuMigrations | cpu-migrations | Number of times the process has migrated to a new CPU |
| ~~PerfPageFaultsMin~~ | page-faults-min | Number of minor page faults. These did not require disk I/O to handle |
| ~~PerfPageFaultsMaj~~ | page-faults-maj | Number of major page faults. These required disk I/O to handle |
| PerfAlignmentFaults | alignment-faults | Number of alignment faults. These happen when unaligned memory accesses happen; the kernel can handle these but it reduces performance. This happens only on some architectures (never on x86). |
| PerfEmulationFaults | emulation-faults | Number of emulation faults. The kernel sometimes traps on unimplemented instructions and emulates them for user space. This can negatively impact performance. |

~~Computed Events~~ (not possible in an "increment" manner)
---
| Name | Description |
| --- | --- |
| ~~PERF_CUSTOM_INSTRUCTIONS_PER_CPU_CYCLE_SCALED~~ | PERF_COUNT_HW_INSTRUCTIONS / PERF_COUNT_HW_CPU_CYCLES |
| ~~PERF_CUSTOM_INSTRUCTIONS_PER_CPU_CYCLE~~ | PERF_COUNT_HW_INSTRUCTIONS / PERF_COUNT_HW_REF_CPU_CYCLES |

And42 committed:
Software events for perf_events metrics

And42 committed:
Removed unused code

And42 committed:
Correct multithread metrics recording

And42 committed:
Fixed logging for events + added hardware event and a couple of custom ones

And42 committed:
Removed map from the perf events

And42 committed:
Moved events' comments to the definitions file

And42 committed:
Paranoid str to int check

And42 committed:
Reformatting code + static initializers

And42 commented:
@alexey-milovidov discussion-ready pull request created 🎉🎉🎉

hodgesds commented on dbms/src/Common/ThreadProfileEvents.cpp:
```diff
@@ -0,0 +1,218 @@
+#include <cerrno>
+#endif
+
+namespace DB {
```
It may be nice to have separate methods (Start, Stop, Reset) to support the `ioctl` options on a profilers like `PERF_EVENT_IOC_RESET`, `PERF_EVENT_IOC_ENABLE` and `PERF_EVENT_IOC_DISABLE`. They sort of exist in the current implementation, but it would be nice if you could have those extra controls.

And42 replied:
Can you please describe how those (Start, Stop, Reset) methods can be called from outside of the ClickHouse process (or otherwise, I don't understand when they are useful within ClickHouse codebase)? Probably, due to the lack of knowledge and/or experience I do not understand the phrase "to support the `ioctl` options on a profilers like `PERF_EVENT_IOC_RESET`, `PERF_EVENT_IOC_ENABLE` and `PERF_EVENT_IOC_DISABLE`" and googling didn't help with that, unfortunately. So, will be glad to hear about examples (or just the same thing but with a bit of explanation) when these methods can be useful

hodgesds replied:
Ideally they would be used when a user updates ClickHouse settings. For example, all the `PERF_COUNT_HW_*` events could have a setting of `perf_hardware_events_enabled` or something similar that would also expose all these metrics via graphite or a system table. Perf counters are increasing while they are enabled (however they can wrap uint64s). The difficulty becomes when you want to restart profiling, if you disable the configuration should the counter be reset to zero or continue from the last value (this is what the ioctls would handle). Then you could have another configuration field such as `perf_hardware_events_reset_on_enable` to control whether an profiler is reset when it is enabled.

alexey-milovidov commented:
@ludv1x Please don't worry about PR assigments by Ivan Blinkov. He has written a script that assigns pull request to random people. We decided that we will remove this script but Ivan has not stopped it yet.

And42 committed:
Merge branch 'master' into perf_events_metrics

# Conflicts:
#	dbms/src/Common/ProfileEvents.cpp

And42 committed:
Fixed styling

And42 committed:
Fixed compilation issues

And42 committed:
Handle `CAP_SYS_ADMIN` permission accordingly

And42 committed:
Replaced some warnings with infos as perf events measurements are optional

And42 committed:
Fixed styling v2

akuzm committed:
boop the CI

akuzm commented:
The performance tests look bad. It seems that initializing these events is really slow: https://sendeyo.com/up/d/ba99cdd31d
(flame graph from the per fest output, you can download it using the link at the end of the report)

alexey-milovidov commented:
@akuzm Is it possible to avoid initialization/deinitialization all the time?

And42 commented:
@alexey-milovidov @akuzm do queries reuse threads from a pool or create new ones when executed? If they reuse it's possible to call `syscall` only when a thread is created and not each time a query starts

alexey-milovidov commented:
@And42 Yes, queries reuse threads from pool.

alexey-milovidov commented:
You can use thread local variable as a flag to check if perf-events are initialized.

And42 committed:
Using the same file descriptors for all counters on the current thread (only one instance of `PerfEventsCounters` can be active at a given time for a thread)

And42 committed:
perf events' values are 64 bit unsigned; so, read them instead of the signed ones

And42 committed:
A bit of reformatting code

And42 commented:
In the last commit I added a logic to create descriptors only once per thread. This added an issue: there can be only one active instance of `PerfEventsCounters` per thread at a given moment. If there is more than a one instance the resulting values are incorrect. I will fix it by not stopping event measurements after a query finishes, but for now just want to see if that's enough to not fail the performance tests

And42 committed:
Fix styling

And42 commented:
Well, one performance test is failing anyway :(
For now the only solution I see if that's critical is to add an additional switch to preferences to enable/disable perf events

alexey-milovidov commented:
> Well, one performance test is failing anyway :(

That's well known flacky test, it fails sporadically. So we can say that performance tests are passed!

akuzm commented:
A couple of interesting files with new metrics from the performance test. You can find them in the performance test output archive (click "test output" at the bottom of the performance test report).  These are the metrics for two of the troublesome queries -- the `visitParam` one often changes performance with no apparent reason, and the `codecs` one always has very big spread.

[SELECT count() FROM zeros(2000000) WHERE NOT ignore(visitParamExtractRaw(materialize(\'{"myparam":{"nested_1":"test_stri.right.metrics.txt](https://github.com/ClickHouse/ClickHouse/files/4544168/SELECT.count.FROM.zeros.2000000.WHERE.NOT.ignore.visitParamExtractRaw.materialize.myparam.nested_1.test_stri.right.metrics.txt)
[INSERT INTO codec_seq_UInt64_DoubleDelta (n) SELECT number FROM system.numbers LIMIT 10000000 SETTINGS max_threads=1.right.metrics.txt](https://github.com/ClickHouse/ClickHouse/files/4544169/INSERT.INTO.codec_seq_UInt64_DoubleDelta.n.SELECT.number.FROM.system.numbers.LIMIT.10000000.SETTINGS.max_threads.1.right.metrics.txt)

These files show the variation of a particular metric over several runs of the same query as [min, median, max]. Some things I noticed there:

* `PERF_CUSTOM_INSTRUCTIONS_PER_CPU_CYCLE_SCALED` looks kinda strange, sometimes it's 1 and sometimes 4636. What might be the reason? The values are from `system.query_log.ProfileEvents`.

* The metric names don't follow the style of our other metrics. Since this is a user-facing feature, maybe we should adopt the names that `perf stat` uses? Like `iTLB-load-misses`. They may be more familiar to users and also look less scary.

* For page faults, we now have `PERF_COUNT_SW_PAGE_FAULTS`, `PERF_COUNT_SW_PAGE_FAULTS_MIN` and `SoftPageFaults`, that apparently have the same values. What is the difference?

akuzm commented:
One thing that bothers me is that the perf event descriptors apparently count toward the open file limit, and it is not unusual for us to have thousands of threads, so we may easily run into the limit. It should be possible to check by running something like `select count(*) from system.zeros max_treads = 4096`. The real number of threads started for a query can be found in `system.query_thread_log`.

And42 commented:
> PERF_CUSTOM_INSTRUCTIONS_PER_CPU_CYCLE_SCALED looks kinda strange, sometimes it's 1 and sometimes 4636. What might be the reason? The values are from system.query_log.ProfileEvents.

Seems like a bug, as `PERF_CUSTOM_INSTRUCTIONS_PER_CPU_CYCLE_SCALED` = `PERF_COUNT_HW_INSTRUCTIONS` / `PERF_COUNT_HW_CPU_CYCLES`. And this value should be `198561861` / `113836367` = `1.74...` according to the log.

> The metric names don't follow the style of our other metrics. Since this is a user-facing feature, maybe we should adopt the names that perf stat uses? Like iTLB-load-misses. They may be more familiar to users and also look less scary.

Definitely will be renamed. The current names are used only on the development stage to easily map them to the names for events in the C++ code.

> For page faults, we now have PERF_COUNT_SW_PAGE_FAULTS, PERF_COUNT_SW_PAGE_FAULTS_MIN and SoftPageFaults, that apparently have the same values. What is the difference?

`PERF_COUNT_SW_PAGE_FAULTS_MIN` = soft page faults = `SoftPageFaults`
`PERF_COUNT_SW_PAGE_FAULTS` = soft + hard page faults
the only difference between `PERF_COUNT_SW_PAGE_FAULTS_MIN` and `SoftPageFaults` is that the first one is taken from perf_events while the second one is from rusage. So, actually seems like there is a reason to leave only one of them and since rusage does not use file descriptors, it's better to remove `PERF_COUNT_SW_PAGE_FAULTS_MIN`, `PERF_COUNT_SW_PAGE_FAULTS_MAJ` and `PERF_COUNT_SW_PAGE_FAULTS`.

> One thing that bothers me is that the perf event descriptors apparently count toward the open file limit ...

Have no idea how to fix it at the moment without sacrificing some part of the measurements but will think about it.

And42 committed:
Merge branch 'master' into perf_events_metrics

# Conflicts:
#	src/Common/ThreadProfileEvents.cpp
#	src/Common/ThreadProfileEvents.h
#	src/Common/ThreadStatus.cpp
#	src/Common/ThreadStatus.h

And42 committed:
Updated modified submodules

And42 committed:
Removed website/images/feathericons

And42 committed:
Added contrib/libcpuid

And42 committed:
Fixed compilation namespace issue and formatting

And42 commented:
@akuzm seems like the best way is to just copy-paste event names from https://github.com/torvalds/linux/blob/master/tools/perf/util/parse-events.c (i.e. `perf stat` names you've suggested). The only problem is perf has GPL-2.0 license. Don't really know how to better integrate the part of its code here without need to relicense ClickHouse under GPL-2.0

alexey-milovidov commented:
It's ok to leave ALL_CAPS event names for simplicity or to invent our own convenient names for them.

And42 committed:
Added running and enabled time for perf metrics

And42 committed:
Renamed events

And42 committed:
Removed custom events as they don't work in an "increment" manner

akuzm commented:
Having a third set of names that is neither API constants nor perf stat names is the worst option.
Don't worry about this now, I'll rename them myself before merging.

And42 committed:
Fixed 0 value of paranoid handling

And42 committed:
Removed unused method

And42 committed:
UInt64 numbers to track active counters instead of memory addresses

And42 committed:
Moved static on top of thread_local

And42 committed:
Changed comments

And42 committed:
Reverted switch case indentation

And42 committed:
Reverted switch case indentation (v2)

And42 committed:
Fixed current counter id check

And42 committed:
Revert "Fixed current counter id check"

This reverts commit f98c53e8

alexey-milovidov commented:
There is a special exception in Linux license that permits to use API of system calls, but I'm not sure that these names are part of API.

And42 commented:
So, for now the remaining issues are:
1. Too many file descriptors overall
2. Multiplexing + scaling in case there are more hardware counters enabled than the number of actual hardware counters units (not an issue by itself, but no ability to avoid it is an issue)

There is no way around to avoid creating a file descriptor for an event if you want to measure it, so, all the solutions can be related either to closing file descriptors when they are not needed or not opening them entirely (and not measuring some of events) in the first place.
Closing descriptors when not needed proved to be a "performance degradation" way as can be seen from the tests; so, seems like it's better to not open them entirely. This can be done if new settings are added to the ClickHouse, i.e. whether to use perf metrics at all and an array/list/set/whatever that contains all the events that a user wants to measure (perf event names could be handy there). This way a user can lower the number of opened file descriptors and additionally can avoid multiplexing (by disabling some hardware events).
Another way (or an additional measure) to deal with open file descriptors is to set some threshold for a number of opened file descriptors, i.e. if the current number of descriptors is more than 70% of a soft limit, do not open new ones and log warnings to not break other CH logic dependent on descriptors.

IMHO, the best way for now is to add two settings (`enable_perf_events`: `defaults to false` and `perf_events_list`: `defaults to all events`) and a threshold.

And42 committed:
Fixed warnings

akuzm commented:
> There is a special exception in Linux license that permits to use API of system calls, but I'm not sure that these names are part of API.

I doubt it's possible to have any kind of license for something like 'dTLB-store-misses', these are just common words. And we won't use any of their code, we just need to use the same names for the same counters.

akuzm commented:
> IMHO, the best way for now is to add two settings (`enable_perf_events`: `defaults to false` and `perf_events_list`: `defaults to all events`) and a threshold.

Sounds good to me.

And42 committed:
Added file descriptors threshold

And42 committed:
typedef > using

And42 committed:
Settings for measuring only particular metrics via perf events

And42 committed:
Enable perf measurements by default for performance tests

And42 committed:
Space

And42 committed:
Removed useless comment

And42 committed:
Fixed compilation on non linux devices

And42 committed:
Fixed naming for lambda

And42 committed:
Validating file descriptors for external modification

And42 committed:
Logging perf event settings name when event is not supported

And42 committed:
Logging warnings instead of infos and always as the user can select
what metrics he wants to measure and perf measurements will be disabled by default

And42 committed:
More precise number of total events

And42 committed:
Renamed `counters_id` to `latest_counters_id` to be more specific about what this variable holds

And42 committed:
Reverted `ctime` > `sys/time.h` include just to not show it in the pull request

And42 committed:
Added `cpu-clock` to the list of supported events as it is included in perf and can be disabled if needed

And42 committed:
Removed perf page faults as they already exist in rusage

And42 committed:
Perf metrics disabled by default

And42 commented:
@akuzm @alexey-milovidov Is there any way to suppress a specific style warning? Having double (and even tripple whitespaces) whitespaces in the comment as an example of a valid value: https://clickhouse-test-reports.s3.yandex.net/9545/e3b756e65f969fa0be5f69b71eb501da2e6fbfe6/style_check/output.txt

alexey-milovidov commented:
@And42 Two ways:
- use at least 4 whitespaces;
- rephrase this comment;

And42 committed:
1 or 4 spaces in example values

akuzm committed:
Enable perf event metrics in performance test

qoega commented:
we get error in ya make build:
```
ld: error: undefined symbol: __cxa_thread_atexit
>>> referenced by ThreadProfileEvents.cpp:166 (/-S/clickhouse/src/Common/ThreadProfileEvents.cpp:166)
>>>               ThreadProfileEvents.cpp.o:(DB::PerfEventsCounters::processThreadLocalChanges(std::__y1::basic_string<char, std::__y1::char_traits<char>, std::__y1::allocator<char> > const&)) in archive clickhouse/src/Common/libclickhouse-src-Common.a
```
@akuzm We need help to fix build.

And42 commented:
@akuzm @qoega as far as I can tell from the log, the issue is likely to be related to the "libc++abi" version used in the build process. This is mentioned in the boost sources, for example: https://www.boost.org/doc/libs/1_73_0/boost/config/stdlib/libcpp.hpp. And according to [this](https://github.com/intel/linux-sgx/issues/454#issuecomment-539382295) it's related to the custom destructor of the `thread_local` variable.
Seems like there are two ways to fix this:
1. On the build system side you can update the compiler version that is used, so that it uses the version of libc++abi that contains this method (though, I don't really know what exact version contains it)
2. On the code side I can replace `thread_local` for this particular variable to `boost::thread_specific_ptr` that should work in all scenarios

I can't make the decision on my own as I don't know the requirements and the possibility of the 1'st variant. The 2'nd variant seems ok but ideally there should be added a compile time check (lint/style check/etc) checking whether `thread_local` is used on a variable with a custom destructor as this issue can arise again in the future otherwise

alexey-milovidov commented:
@qoega This method exists in C++ ABI since C++11 to support thread_local variables, but Arcadia build has specifically broken version of libc++ that lack this method.

And42 committed:
Fix for values scaling due to multiplexing and `time_enabled` + `time_running` deltas calculation

And42 committed:
Fix for processing deltas in case an instance of `PerfEventCounters` is different from the previous one

And42 commented:
Merged within pull request: https://github.com/ClickHouse/ClickHouse/pull/11226

amosbird commented:
Do we have estimated number of overheads when enabling `metrics_perf_events_enabled`?

alexey-milovidov commented:
No, but it should be rather small, as we enable it in our perf tests.
